File: src/pann/NetworkModels/MultilayerPerceptron.cpp
//MultilayerPerceptron.cpp

#include "Core/Net.h"
#include "Core/Weight.h"
#include "NeuronFactories/PyramidalNeuron.h"

#include "MultilayerPerceptron.h"

using std::vector;
using boost::tuple;

namespace pann
{
    NetPtr
    MultilayerPerceptron(vector<tuple<unsigned, TfPtr> > layers)
    {
        NetPtr net(new Net());

        if(layers.size() == 0)
            return net;

        vector<vector<NeuronPtr> > mlp(layers.size());

        //Layers
        //TODO: OpenGlHint
        for(unsigned l = 0; l < layers.size(); ++l)
        {
            TfPtr af = layers[l].get<1>();
            if(!af)
                af = TanH::Instance();

            for(unsigned i = 0; i < layers[l].get<0>(); ++i)
            {
                NeuronPtr n;
                if(l == 0)
                {
                    n = NeuronFactory::PyramidalNeuron(af);
                    net->add_input_neuron(n);
                } else {
                    n = NeuronFactory::PyramidalNeuron(af, WeightPtr(new Weight(1)));
                }

                mlp[l].push_back(n);
            }
        }

        //Connections
        for(unsigned i = 0; i < mlp.size() - 1; i++) //layers
            for(unsigned j = 0; j < mlp[i].size(); j++) //prev layer
                for(unsigned k = 0; k < mlp[i+1].size(); k++) //next layer
                    //Connection from current layer (i) to next (i+1)
                    net->add_connection(mlp[i][j], mlp[i+1][k]);

        return net;
    };

}; //pann

File: pypann/pypann/NetworkModels/MultilayerPerceptron.py
#
# MultilayerPerceptron
#

from ..Core.Weight import Weight
from ..Core.Net import Net
from ..Neurons.PyramidalNeuron import PyramidalNeuron
from ..Neurons.BiasNeuron import BiasNeuron
from ..TF import Linear, Tanh

def multilayer_perceptron(layers, tfs = []):
    net = Net()

    if len(layers) == 0:
        return net

    mlp = []
    for i in range(len(layers)):
        if i > len(tfs) - 1:
            if i == len(layers) - 1:
                tf = Linear()
            else:
                tf = Tanh()
        else:
            tf = tfs[i]

        mlp.append([])
        for j in range(layers[i]):
            n = PyramidalNeuron(tf)
            mlp[i].append(n)
            if i == 0:
                net.add_input_neuron(n)
            else:
                net.connect(BiasNeuron(), n, Weight(1.0))

    for i in range(len(mlp) - 1):
        for j in range(len(mlp[i])):
            for k in range(len(mlp[i+1])):
                net.connect(mlp[i][j], mlp[i+1][k], Weight(1.0))

    return net

